drivers: modem: gsm_ppp: Optimizations and FSM implemetation

ycsin opened the pull request:
Some optimizations, code cleanup and implemented the gsm_ppp to configure in FSM.

ycsin committed:
drivers: modem: gsm: Remove unused flags

The `setup_done` and `mux_setup_done` aren't being used
anywhere in the driver, therefore should be removed.

Signed-off-by: Yong Cong Sin <yongcong.sin@gmail.com>

ycsin committed:
drivers: modem: gsm: Reorder gsm mux logs so that it will be printed

The gsm_configure will return if it fails to perform mux_enable,
therefore the disable part of the log will not be printed.

Signed-off-by: Yong Cong Sin <yongcong.sin@gmail.com>

Update gsm_ppp.c

ycsin committed:
drivers: modem: modem_cmd_handler: Fix missing documentations

Fix missing documentations.

Signed-off-by: Yong Cong Sin <yongcong.sin@gmail.com>

ycsin committed:
drivers: modem: gsm: Recycle retries variable

The attachment and the rssi readout are done in sequence so it
is possible to just recycle the same retries counter.

Signed-off-by: Yong Cong Sin <yongcong.sin@gmail.com>

ycsin committed:
drivers: modem: gsm: Remove unnecessary mux_enabled check

Remove redundant mux_enabled checks, the code execution will
not reach here if mux_enabled is false in the first place.

Signed-off-by: Yong Cong Sin <yongcong.sin@gmail.com>

ycsin committed:
drivers: modem: gsm: Use FSM implementation

Combine the setup_state and booleans into a FSM that is simpler
and easier to understand the status of modem.

Signed-off-by: Yong Cong Sin <yongcong.sin@gmail.com>

ycsin committed:
drivers: modem: gsm: Customize error logs to ease debugging

Customize the error logs in each connection finalization steps
so that it is easier to trace the error.

Signed-off-by: Yong Cong Sin <yongcong.sin@gmail.com>

ycsin commented:
@jukkar @rlubos @endian-benjamin 
I have created this PR for the FSM and stuffs.

I've added a comment at the `mux_setup` when it [fails](https://github.com/zephyrproject-rtos/zephyr/pull/38596/files#diff-49a455a63fb862468ab36610ec19ec3fb7fa65b9f1981831865fd9fc064c5df3R920). It seems reasonable to to reschedule the `gsm_configure_work` at `fail:` so that it retries the same step again, but I'm not sure if retrying the same step helps when it fails.

@jukkar is this something that you had in mind when creating this driver?

ycsin commented on drivers/modem/gsm_ppp.c:
```diff
@@ -913,13 +909,18 @@ static void mux_setup(struct k_work *work)
 
 		gsm_finalize_connection(gsm);
```
Need to handle gsm->state set in `gsm_finalize_connection()` and fails, easier if change `gsm_finalize_connection` into a work.

jukkar commented:
> I need something like that in my application to know what's the current GSM state of the modem.

@bbilas What kind of use cases you have for this, why application would need to know the internal state of the GSM modem?

bbilas commented:
@jukkar One use case is to know and inform the user about the exact gsm modem state (gsm is attaching, attached etc.). The second one is to inform the user that the connection is established (but ok - it can be done using net_mgmt_event). One of the other requirements is to get modem information such as imei, imsi and rssi but that would need a bit more work because as I see modem_context struct isn't public so currently there is no easy way to get those data in the application. Is there any chance to make it public to use e.g modem_context_from_id function to retrieve that information easily?

ycsin committed:
drivers: modem: gsm: Convert gsm_finalize_connection to work

Convert gsm_finalize_connection into a work so that the caller
work won't have to run again when gsm_finalize_connection
reschedule gsm_configure_work on error.

Signed-off-by: Yong Cong Sin <yongcong.sin@gmail.com>

ycsin committed:
drivers: modem: gsm: Go to next state if mux_attach successfully

Currently the mux_setup set the state to the next one after
uart_mux_alloc is successful even if the mux_setup fails which
can be a problem.

Set the state after both mux_setup & uart_mux_alloc are
successful.

Signed-off-by: Yong Cong Sin <yongcong.sin@gmail.com>

ycsin committed:
drivers: modem: gsm: Restart mux_setup work on failure

Previously, nothing is being done if something fails inside
mux_setup work. This patches it to reschedule itself on
error.

Signed-off-by: Yong Cong Sin <yongcong.sin@gmail.com>

ycsin commented:
I've added few more commits to this pr that focus on error recovery, it is marked as draft as I haven't test the code yet.

jukkar reviewed (commented):
> Would love your comments on the mux_setup work's error handling

Any specific commit you wish to get comments?

ycsin committed:
drivers: modem: gsm: Convert k_work to k_work_delayable

Convert k_work to k_work_delayable before passing it into
CONTAINER_OF macro.

Signed-off-by: Yong Cong Sin <yongcong.sin@gmail.com>

mniestroj reviewed (commented):
Would be good to split code and documentation cleanup into separate PR, so that it simplifies review process and can move this batch a bit forward.

tautologyclub commented on drivers/modem/gsm_ppp.c:
```diff
@@ -996,7 +999,8 @@ static void gsm_configure(struct k_work *work)
 		return;
 	}
 
-	gsm_finalize_connection(gsm);
+	k_work_init_delayable(&gsm->gsm_configure_work, gsm_finalize_connection);
+	(void)k_work_reschedule(&gsm->gsm_configure_work, K_NO_WAIT);
```
@mniestroj None of those points are pertinent to this PR in particular. If we're truly abusing work queues there should be a bug filed.

tautologyclub commented on drivers/modem/gsm_ppp.c:
```diff
@@ -1129,8 +1147,11 @@ static int gsm_init(const struct device *dev)
+
 	if (IS_ENABLED(CONFIG_GSM_PPP_AUTOSTART)) {
-		gsm_ppp_start(dev);
+		(void)gsm_ppp_start(dev);
```
seems like you can return `gsm_ppp_start(dev)` here

tautologyclub commented:
Personally I like the changes so far and would suggest not extending the scope of the PR further right now. Most of the objections right now seems to not be introduced by this PR.

tautologyclub commented:
> > > ```c
> > > /* FIXME: The driver can potentially stuck here retrying forever */
> > > ```
> > 
> > 
> > We should not wait forever, but have some timeout so that it is possible to act on that.
> 
> How about a `MAX_RETRIES` define or Kconfig before it just give up and do nothing? I think @bbilas is planning for something that can pass that info out of driver so that the application can handle it

Ideally `gsm_ppp_stop()` should make sure the loop is aborted. If so, an application can just wait for the appropriate net_event and if that has not arrived after X seconds, they can interpret that as a failure and call `gsm_ppp_stop()`. 

This may not be the theoretically optimal solution, but if it's viable at all then I'd recommend handling it in another PR.

tautologyclub commented:
Btw `gsm_finalize_connection()` should probably start by checking if state == stopped and just return if so. That way you can rely on `gsm_ppp_stop()` to actually cancel the work item. Without this, it could get cancelled and then rescheduled by the work handler itself after `gsm_ppp_stop()` was called.

ycsin commented on drivers/modem/gsm_ppp.c:
```diff
@@ -581,30 +584,32 @@ static void rssi_handler(struct k_work *work)
+					     gsm_configure_work);
 
 	/* If already attached, jump right to RSSI readout */
-	if (gsm->attached) {
```
I might split these into different modem states and do it in different work item so that it is cleanner and doesn't have to jump around at the top, what do you all think?

@jukkar @bbilas @tautologyclub @mniestroj @endian-benjamin @rlubos

tautologyclub replied:
My preference would be to get this merged. Make a new PR with further improvements

tautologyclub commented:
@ycsin Seems like you sort of haven't decided on the scope of this PR yet. I get the feeling this is "on its way in", which means I'm hesitant about starting work on other gsm_ppp patches. Do you mind trying to finish this up?

ycsin commented:
> @ycsin Seems like you sort of haven't decided on the scope of this PR yet.

I do not plan to increase the scope of this PR. I just don't have time to get my custom board compile on this branch yet. Worked on two existing PRs today and personally I'd like to do this next, but I have other things to attend to for my job this week that unfortunately takes precedence over this PR
